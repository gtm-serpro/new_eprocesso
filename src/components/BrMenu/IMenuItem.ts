interface IMenuItem {
  label?: string;
  link?: string;
  submenu?: IMenuItem[];
  icon?: string;
  divider?: boolean;
  expanded?: boolean;
  active?: boolean;
  danger?: boolean;
  keepMenuActive?: boolean;
  onClick?: () => void;
}

export default IMenuItem;
