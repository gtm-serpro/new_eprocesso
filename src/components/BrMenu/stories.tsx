import { Meta, Story } from "@storybook/react";

import BrMenu, { MenuProps } from ".";
import BrButton from "../BrButton";

export default {
  title: "Navegação/Menu",
  component: BrMenu,
  args: {
    closable: true,
    id: "main-navigation",
    systemName: "Nome da aplicação",
    systemLogoUrl: "https://www.gov.br/ds/assets/img/govbr-logo.png",
    data: [
      {
        label: "Página Inicial",
        icon: "home",
        onClick: () => alert("Página Inicial"),
        divider: true,
      },
      {
        label: "Folder",
        icon: "calendar",
        submenu: [
          {
            label: "Sub Folder 1",
            onClick: () => alert("Sub Folder 1"),
            icon: "moon",
          },
          {
            label: "Sub Folder 2",
            icon: "sun",
            submenu: [
              {
                label: "Sub Sub Folder 1",
                icon: "wifi",
                onClick: () => alert("Sub Sub Folder 1"),
              },
            ],
          },
        ],
      },
    ],
    logos: [
      {
        src: "https://www.gov.br/ds/assets/img/govbr-logo.png",
        alt: "Logo 01",
      },
      {
        src: "https://www.gov.br/ds/assets/img/govbr-logo.png",
        alt: "Logo 02",
      },
    ],
    externalLinks: [
      {
        link: "https://google.com/",
        label: "Link externo 01",
      },
      {
        link: "https://google.com/",
        label: "Link externo 02",
      },
    ],
    socialNetworks: [
      {
        link: "#",
        icon: "fab fa-facebook-f",
        name: "Facebook",
      },
      {
        link: "#",
        icon: "fab fa-twitter",
        name: "Twitter",
      },
      {
        link: "#",
        icon: "fab fa-linkedin-in",
        name: "Linkedin",
      },
      {
        link: "#",
        icon: "fab fa-whatsapp",
        name: "Whatsapp",
      },
    ],
    info: (
      <div className="text-center text-down-01">
        Todo o conteúdo deste site está publicado sob a licença{" "}
        <strong>Creative Commons Atribuição-SemDerivações 3.0</strong>
      </div>
    ),
  },
} as Meta;

const Template: Story<MenuProps> = (args) => (
  <>
    <BrButton data-target="#main-navigation" icon="bars" size="small">
      Ativar menu
    </BrButton>
    <BrMenu {...args} />
  </>
);

export const Normal = Template.bind({});

export const Push = Template.bind({});
Push.args = { type: "push", closable: undefined };

export const Contextual = Template.bind({});
Contextual.args = { type: "contextual", contextualTriggerTitle: "Menu contextual" };
