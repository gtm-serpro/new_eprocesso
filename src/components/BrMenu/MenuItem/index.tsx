/* eslint-disable no-script-url */
import { useWindowWidth } from "@/hooks/useWindowWidth";
import classNames from "classnames";
import React, { useEffect, useMemo, useRef } from "react";
import CustomTag from "../../../common/CustomTag";
import { useMtProps } from "../../../hooks/useMtProps";
import { useSpreadProps } from "../../../hooks/useSpreadProps";
import useUniqueId from "../../../hooks/useUniqueId";
import IMtProps from "../../../types/IMtProps";
import IMenuItem from "../IMenuItem";

interface MenuItemProps extends Omit<React.HTMLAttributes<HTMLDivElement>, "expanded">, IMtProps {
  label?: string;
  href?: string;
  icon?: string;
  submenu?: IMenuItem[];
  level?: number;
  divider?: boolean;
  expanded?: boolean;
  active?: boolean;
  danger?: boolean;
  keepMenuActive?: boolean;
  onClick?: () => void;
  menuType?: "normal" | "push" | "contextual";
  refMenuObject: React.MutableRefObject<any>;
}

type RefType = HTMLDivElement | HTMLAnchorElement;

const hasItemActiveRecursive = (items?: IMenuItem[]): boolean => {
  return (
    !!items?.length && items?.some((item) => item?.active || hasItemActiveRecursive(item?.submenu))
  );
};

const getParentFolder = (element?: Element | null) =>
  element?.parentElement?.closest(".side-menu") ||
  element?.parentElement?.closest(".menu-folder") ||
  element?.parentElement?.closest(".menu-body");

const MenuItem = React.forwardRef<HTMLDivElement, MenuItemProps>(
  (
    {
      className,
      children,
      id,
      label,
      level = 2,
      href,
      icon,
      submenu,
      divider = false,
      expanded = false,
      active = false,
      danger = false,
      keepMenuActive = false,
      onClick,
      menuType = "normal",
      refMenuObject,
      ...props
    },
    ref
  ) => {
    const fid = useUniqueId(id, "menuitem_____");
    const mtProps = useMtProps(props);
    const spreadProps = useSpreadProps(props);

    const itemRef = (ref ?? useRef<RefType>(null)) as React.MutableRefObject<RefType>;
    const itemElement = itemRef.current;
    const hasActive = useMemo(() => hasItemActiveRecursive(submenu), [submenu]);
    const windowWidth = useWindowWidth();

    useEffect(() => {
      if (itemElement && active) {
        itemElement.removeAttribute("hidden");
        itemElement.classList.add("active");

        const folder = getParentFolder(itemElement);
        const siblings = Array.from(
          (folder?.classList.contains("menu-folder")
            ? getParentFolder(folder)
            : folder
          )?.querySelectorAll(".menu-item") ?? []
        );

        siblings?.forEach((sibling) => {
          if (sibling !== itemElement) {
            sibling.removeAttribute("hidden");
            sibling.classList.remove("active");
          }
        });

        let foldersInTree: Element[] = [];
        let currentFolder = folder;

        while (currentFolder) {
          if (folder?.classList.contains("side-menu") && folder !== currentFolder) {
            currentFolder.querySelectorAll(".menu-item").forEach((el) => {
              if (!siblings.includes(el)) {
                el.setAttribute("hidden", "");
              }
            });
          }
          foldersInTree.push(currentFolder);
          currentFolder = getParentFolder(currentFolder);
        }

        const otherFolders = itemElement
          .closest(".menu-body")
          ?.querySelectorAll(".menu-folder, .side-menu");

        otherFolders?.forEach((otherFolder) => {
          if (foldersInTree.includes(otherFolder)) {
            otherFolder.classList.add("active");
          } else {
            otherFolder.classList.remove("active");
          }
        });
      }
    }, [itemElement, active]);

    return (
      <CustomTag
        {...(!!submenu?.length && level === 2 && { tagName: "div" })}
        className={classNames(!!submenu?.length && "menu-folder drop-menu", hasActive && "active")}
      >
        <CustomTag
          ref={itemRef}
          id={fid}
          tagName={expanded ? "div" : "a"}
          className={classNames(
            "menu-item",
            divider && "divider",
            className,
            ...mtProps,
            active && "active",
            danger && "text-danger",
            level > 2 && hasActive && "side-menu-init-open"
          )}
          {...(href && !expanded && { href: href })}
          {...(!!submenu?.length &&
            !expanded && { href: "#", onClick: (e: Event) => e.preventDefault() })}
          {...(onClick && {
            href: href ?? "#",
            onClick: (e: Event) => {
              e.preventDefault();
              e.stopPropagation();
              onClick?.();
              if (
                !keepMenuActive &&
                !submenu?.length &&
                (["normal", "contextual"].includes(menuType) || windowWidth < 576)
              ) {
                refMenuObject?.current?._closeMenu();
              }
            },
          })}
          {...spreadProps}
        >
          <span className="icon">
            {icon && (
              <i className={icon.includes("fas ") ? icon : `fas fa-${icon}`} aria-hidden="true"></i>
            )}
          </span>
          <span className="content">{label}</span>
        </CustomTag>

        {!!submenu?.length && (
          <ul>
            {submenu.map((item, index) => (
              <li key={index}>
                <MenuItem
                  href={item.link}
                  icon={item.icon}
                  label={item.label}
                  submenu={item.submenu}
                  level={level + 1}
                  divider={item.divider}
                  expanded={item.expanded}
                  active={item.active}
                  danger={item.danger}
                  keepMenuActive={item.keepMenuActive}
                  onClick={item?.onClick}
                  menuType={menuType}
                  refMenuObject={refMenuObject}
                />
              </li>
            ))}
          </ul>
        )}
        {children}
      </CustomTag>
    );
  }
);

export default MenuItem;
