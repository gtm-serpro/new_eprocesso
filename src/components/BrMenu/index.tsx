import { getBreakpoints } from "@/helpers/styles";
import { Breakpoints } from "@/types/layout";
import { BRMenu } from "@govbr-ds/core";
import classNames from "classnames";
import React, { useEffect, useMemo, useRef } from "react";
import useCommonProperties from "../../hooks/useCommonProperties";
import { useMtProps } from "../../hooks/useMtProps";
import { useSpreadProps } from "../../hooks/useSpreadProps";
import useUniqueId from "../../hooks/useUniqueId";
import IMtProps from "../../types/IMtProps";
import IMenuItem from "./IMenuItem";
import IMenuLink from "./IMenuLink";
import IMenuLogo from "./IMenuLogo";
import ISocialNetwork from "./ISocialNetwork";
import MenuItem from "./MenuItem";

export interface MenuProps
  extends React.HTMLAttributes<HTMLDivElement>,
    IMtProps,
    Omit<Breakpoints, "xs"> {
  systemLogoUrl?: string;
  systemName?: string;
  contextualTriggerTitle?: string;
  data: IMenuItem[];
  logos?: IMenuLogo[];
  externalLinks?: IMenuLink[];
  socialNetworks?: ISocialNetwork[];
  info?: React.ReactNode;
  type?: "normal" | "push" | "contextual";
  startActive?: boolean;
  shadow?: boolean;
  closable?: boolean;
  density?: "small" | "normal" | "large";
}

export interface MenuRef extends HTMLDivElement {
  element: HTMLDivElement;
}

const BrMenu = React.forwardRef<MenuRef, MenuProps>(
  (
    {
      className,
      children,
      id,
      data,
      logos,
      externalLinks,
      socialNetworks,
      info,
      systemLogoUrl,
      systemName,
      contextualTriggerTitle,
      type = "normal",
      density = "normal",
      sm,
      md = 4,
      lg = 3,
      xl = 2,
      startActive,
      closable = false,
      shadow,
      ...props
    },
    ref
  ) => {
    const fid = useUniqueId(id, "menu_____");
    const mtProps = useMtProps(props);

    if (type === "normal" || type === "contextual") {
      mtProps.forEach((prop, index) => {
        if (prop && prop.startsWith("m")) {
          delete mtProps[index];
        }
      });
    }
    const spreadProps = useSpreadProps(props);
    const breakpoints = useMemo(() => getBreakpoints({ sm, md, lg, xl }), [sm, md, lg, xl]);
    const refObject = useRef<typeof BRMenu>();
    const refDiv = useRef<HTMLDivElement>(null);
    const refContextualTrigger = useRef<HTMLButtonElement>(null);

    useCommonProperties<MenuRef>(ref, refDiv, {
      get element() {
        return refDiv.current;
      },
    });

    useEffect(() => {
      if (refDiv.current) {
        if (!refObject.current) {
          refObject.current = new BRMenu("br-menu", refDiv.current);
        }

        refObject.current.component = refDiv.current;
        refObject.current.breakpoints = Object.keys(breakpoints);

        refObject.current._setBreakpoints();
        refObject.current._setDropMenu();
        refObject.current._setSideMenu();

        if (type === "contextual") {
          refDiv.current
            .querySelector(".menu-panel")
            ?.classList.remove(...Object.keys(breakpoints));

          if (refContextualTrigger.current) {
            if (!refContextualTrigger.current.hasAttribute("data-click-listener")) {
              refContextualTrigger.current.addEventListener("click", () => {
                if (refDiv.current?.classList.contains("active")) {
                  refObject.current._closeMenu();
                } else {
                  refObject.current._openMenu();
                }
              });

              refContextualTrigger.current?.setAttribute("data-click-listener", "");
            }
          }

          const triggerHeight = refContextualTrigger.current?.offsetHeight;
          document.body.setAttribute(
            "style",
            `padding-bottom: ${triggerHeight}px; margin-bottom: 0 !important;`
          );
        } else {
          document.body.removeAttribute("style");
        }

        return () => {
          document.body.removeAttribute("style");
        };
      }
    }, [
      refDiv.current,
      refContextualTrigger.current,
      refObject.current,
      data,
      type,
      startActive,
      breakpoints,
    ]);

    useEffect(() => {
      document.querySelectorAll(".side-menu-init-open").forEach((el) => {
        const button = el as HTMLButtonElement;
        button.click();
        button.blur();
        button.classList.remove("side-menu-init-open");
      });
    }, []);

    return (
      <>
        {type === "contextual" && (
          <style>
            {`@media (min-width: 992px) {
  .br-menu.contextual.active,
  .br-menu.contextual .menu-trigger {
    position: fixed;
  }
}`}
          </style>
        )}
        <div
          data-breakpoints={Object.keys(breakpoints).join(" ")}
          ref={refDiv}
          id={fid}
          className={classNames(
            "br-menu",
            type === "push" && "push",
            type === "contextual" && "contextual",
            density === "small" && "small",
            density === "large" && "large",
            startActive && "active",
            startActive && type === "push" && "px-0",
            startActive && type === "push" && breakpoints,
            className,
            ...mtProps
          )}
          {...spreadProps}
        >
          {type === "contextual" && (
            <div className="menu-trigger">
              <button
                ref={refContextualTrigger}
                className="br-button primary block py-4"
                type="button"
              >
                {contextualTriggerTitle && <span className="mr-1">{contextualTriggerTitle}</span>}
                <i className="fas fa-chevron-up ml-5" aria-hidden="true"></i>
              </button>
            </div>
          )}

          <div
            className={classNames(
              "menu-container",
              type === "contextual" && shadow && "shadow-lg-up"
            )}
          >
            <div
              className={classNames(
                "menu-panel",
                shadow && "h-auto position-static shadow-lg-right",
                type === "normal" && breakpoints
              )}
            >
              {(systemLogoUrl || systemName || closable) && (
                <div className="menu-header">
                  <div className="menu-title">
                    {systemLogoUrl && <img src={systemLogoUrl} alt={systemName} />}
                    {systemLogoUrl && systemName && (
                      <span className="br-divider vertical mx-2"></span>
                    )}
                    {systemName && <span>{systemName}</span>}
                  </div>
                  {closable && (
                    <div className="menu-close">
                      <button
                        className="br-button circle"
                        type="button"
                        aria-label="Fechar o menu"
                        data-dismiss="menu"
                      >
                        <i className="fas fa-times" aria-hidden="true"></i>
                      </button>
                    </div>
                  )}
                </div>
              )}
              <div className="menu-body">
                {!!data?.length && (
                  <>
                    {data.map((item, index) => (
                      <MenuItem
                        key={index}
                        href={item.link}
                        icon={item.icon}
                        label={item.label}
                        submenu={item.submenu}
                        level={2}
                        divider={item.divider}
                        expanded={item.expanded}
                        active={item.active}
                        danger={item.danger}
                        keepMenuActive={item.keepMenuActive}
                        onClick={item?.onClick}
                        menuType={type}
                        refMenuObject={refObject}
                      />
                    ))}
                  </>
                )}
              </div>
              <div className="menu-footer">
                {logos && (
                  <div className="menu-logos">
                    {logos.map((logo, index) => (
                      <img key={index} src={logo.src} alt={logo.alt} />
                    ))}
                  </div>
                )}
                {externalLinks && (
                  <div className="menu-links">
                    {externalLinks.map((externalLink, index) => (
                      <a key={index} href={externalLink.link}>
                        <span className="mr-1">{externalLink.label}</span>
                        <i className="fas fa-external-link-square-alt" aria-hidden="true"></i>
                      </a>
                    ))}
                  </div>
                )}
                {socialNetworks && (
                  <div className="menu-social">
                    <div className="text-semi-bold mb-1">Redes Sociais</div>
                    <div className="sharegroup">
                      {socialNetworks.map((socialNetwork, index) => (
                        <div key={index} className="share">
                          <a
                            className="br-button circle"
                            href={socialNetwork.link}
                            aria-label={`Compartilhar por ${socialNetwork.name}`}
                          >
                            <i className={socialNetwork.icon} aria-hidden="true"></i>
                          </a>
                        </div>
                      ))}
                    </div>
                  </div>
                )}
                {info && <div className="menu-info">{info}</div>}
              </div>
            </div>
            <div className="menu-scrim" data-dismiss="menu" tabIndex={0} />
          </div>
          {children}
        </div>
      </>
    );
  }
);

export default BrMenu;
