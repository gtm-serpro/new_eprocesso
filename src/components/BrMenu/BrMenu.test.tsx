import { render, screen } from "@testing-library/react";

import BrMenu from ".";

describe("<BrMenu />", () => {
  it("deveria renderizar com a classe br-menu", () => {
    const { container } = render(
      <BrMenu systemName="Título" data={[{ label: "Item 1" }, { label: "Item 2" }]} />
    );

    expect(container.firstChild).toHaveClass("br-menu");

    expect(container.firstChild).toMatchSnapshot();
  });

  it("deveria renderizar o texto do nome do sistema.", () => {
    const { container } = render(
      <BrMenu systemName="Título" data={[{ label: "Item 1" }, { label: "Item 2" }]} />
    );

    expect(screen.getByText(/Título/i)).toBeInTheDocument();
  });

  it("deveria renderizar com a classe contextual", () => {
    const { container } = render(
      <BrMenu
        type="contextual"
        systemName="Título"
        data={[{ label: "Item 1" }, { label: "Item 2" }]}
      />
    );

    expect(container.lastChild).toHaveClass("contextual");
  });

  it("deveria renderizar com a classe push", () => {
    const { container } = render(
      <BrMenu type="push" systemName="Título" data={[{ label: "Item 1" }, { label: "Item 2" }]} />
    );

    expect(container.firstChild).toHaveClass("push");
  });
});
