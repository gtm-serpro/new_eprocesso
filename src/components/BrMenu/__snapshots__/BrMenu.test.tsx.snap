// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrMenu /> deveria renderizar com a classe br-menu 1`] = `
<div
  class="br-menu"
  data-breakpoints="col-md-4 col-lg-3 col-xl-2"
  id="menu_____1"
>
  <div
    class="menu-container"
  >
    <div
      class="menu-panel col-md-4 col-lg-3 col-xl-2"
    >
      <div
        class="menu-header"
      >
        <div
          class="menu-title"
        >
          <span>
            TÃ­tulo
          </span>
        </div>
      </div>
      <div
        class="menu-body"
        role="tree"
      >
        <a
          class="menu-item"
          id="menuitem_____2"
        >
          <span
            class="icon"
          />
          <span
            class="content"
          >
            Item 1
          </span>
        </a>
        <a
          class="menu-item"
          id="menuitem_____3"
        >
          <span
            class="icon"
          />
          <span
            class="content"
          >
            Item 2
          </span>
        </a>
      </div>
      <div
        class="menu-footer"
      />
    </div>
    <div
      class="menu-scrim"
      data-dismiss="menu"
      tabindex="0"
    />
  </div>
</div>
`;
